src/Compilers/CSharp/Portable/BoundTree/DecisionTree.cs

```diff
@@ -0,0 +1,916 @@
+                result = byType.WhenNull = makeDecision(byType.Expression, byType.Type);
+                result.Parent = byType;
+            }
+            else
+            {
+                result = Add(byType.WhenNull, makeDecision);
+            }
+            if (byType.WhenNull.MatchIsComplete && NonNullHandled(byType))
+            {
+                byType.MatchIsComplete = true;
+            }
+            return result;
+        }
+
+        private bool NonNullHandled(DecisionTree.ByType byType)
+        {
+            var inputType = byType.Type.StrippedType();
+            foreach (var td in byType.TypeAndDecision)
+            {
+                var type = td.Key;
+                var decision = td.Value;
+                if (_conversions.ExpressionOfTypeMatchesPatternType(inputType, type) == true &&
+                    decision.MatchIsComplete)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private DecisionTree AddByNull(DecisionTree.Guarded guarded, DecisionMaker makeDecision)
+        {
+            if (guarded.Default == null)
+            {
+                guarded.Default = new DecisionTree.ByType(guarded.Expression, guarded.Type);
+                guarded.Default.Parent = guarded;
+            }
+
+            var result = AddByNull(guarded.Default, makeDecision);
+            if (guarded.Default.MatchIsComplete) guarded.MatchIsComplete = true;
+            return result;
+        }
+
+        private DecisionTree Add(DecisionTree decision, DecisionMaker makeDecision)
+        {
+            if (decision.MatchIsComplete) return null;
+            switch (decision.Kind)
+            {
+                case DecisionTree.DecisionKind.ByType:
+                    return Add((DecisionTree.ByType)decision, makeDecision);
+                case DecisionTree.DecisionKind.ByValue:
+                    return Add((DecisionTree.ByValue)decision, makeDecision);
+                case DecisionTree.DecisionKind.Guarded:
+                    return Add((DecisionTree.Guarded)decision, makeDecision);
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(decision.Kind);
+            }
+        }
+
+        private DecisionTree Add(DecisionTree.Guarded guarded, DecisionMaker makeDecision)
+        {
+            if (guarded.Default != null)
+            {
+                if (guarded.Default.MatchIsComplete) return null;
+                var result = Add(guarded.Default, makeDecision);
+                if (guarded.Default.MatchIsComplete) guarded.MatchIsComplete = true;
+                return result;
+            }
+            else
+            {
+                var result = guarded.Default = makeDecision(guarded.Expression, guarded.Type);
+                guarded.Default.Parent = guarded;
+                if (guarded.Default.MatchIsComplete) guarded.MatchIsComplete = true;
+                return result;
```

Comment:
It looks like two last statements are identical for both branches and could be moved out of them.